### PR TITLE
Fix shields 'Block 3rd party cookies' option

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Fixes https://github.com/brave/browser-android-tabs/issues/960, https://github.com/brave/browser-android-tabs/issues/963, https://github.com/brave/browser-android-tabs/issues/964 .

### CODE REVIEW

File: chrome/browser/net/blockers/shield_exceptions.cc [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#include "chrome/browser/net/blockers/shield_exceptions.h"
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+namespace net {
+namespace blockers {
+
+bool Is3pCookiesAllowedByWhiteList(const std::string& host) {
```
Comment (bbondy) [inline chrome/browser/net/blockers/shield_exceptions.cc]: Could we pass in the 1p and 3p hosts here to determine via code like this?
https://github.com/brave/brave-core/blob/master/common/shield_exceptions.cc#L110

Comment (AlexeyBarabash) [reply chrome/browser/net/blockers/shield_exceptions.cc]: yes, sure, I am modifying this.

File: components/content_settings/core/browser/cookie_settings.cc [hunk 2]
```diff
@@ -125,7 +126,8 @@ void CookieSettings::GetCookieSetting(const GURL& url,
   // by default, apply CONTENT_SETTING_BLOCKED.
   bool block_third = info.primary_pattern.MatchesAllHosts() &&
                      info.secondary_pattern.MatchesAllHosts() &&
-                     ShouldBlockThirdPartyCookies(first_party_url) &&
+                     ShouldBlockThirdPartyCookies(!first_party_url.is_empty() ?
+                         first_party_url : url) &&
```
Comment (diracdeltas) [inline components/content_settings/core/browser/cookie_settings.cc]: when is first_party_url empty? this seems like it should just check `first_party_url`, not fall back to `url`

File: chrome/browser/net/blockers/shield_exceptions.cc (hunk 1 above)
Comment (AlexeyBarabash) [reply chrome/browser/net/blockers/shield_exceptions.cc]: For now the working white list function looks:
```
bool IsWhitelistedCookieExeption2(const GURL& firstPartyOrigin,
    const GURL& subresourceUrl) {
  // Note that there's already an exception for TLD+1, so don't add those here.
  // Check with the security team before adding exceptions.
  if (firstPartyOrigin.is_empty()) {
    static std::vector<URLPattern> htcs =
    {
      URLPattern(URLPattern::SCHEME_ALL, "https://www.htcsense.com/assets/*"),
      URLPattern(URLPattern::SCHEME_ALL, "https://www.htcsense.com/htcaccount.js"),
      URLPattern(URLPattern::SCHEME_ALL, "https://www.htcsense.com/htcaccount/*"),
    };

    bool ret = std::any_of(htcs.begin(), htcs.end(),
        [&subresourceUrl](const URLPattern& pattern) {
          return pattern.MatchesURL(subresourceUrl);
        });
    return ret;
  }
  static std::vector<std::pair<URLPattern, std::vector<URLPattern>>> whitelist_patterns = {
    {
      URLPattern(URLPattern::SCHEME_ALL, "https://estore.htc.com/tw/buy/zh-TW/shop/LogonForm?*"),
      {
        URLPattern(URLPattern::SCHEME_ALL, "https://www.htcsense.com/htcaccount/remotemethod.html?*"),
        URLPattern(URLPattern::SCHEME_ALL, "https://www.htcsense.com/htcaccount.js"),
        URLPattern(URLPattern::SCHEME_ALL, "https://www.htcsense.com/assets/htcaccount/*"),
        URLPattern(URLPattern::SCHEME_ALL, "https://www.htcsense.com/$WS$/Services/OAuth/Authorize?*"),
      }
    },
  };

  const auto i = std::find_if(whitelist_patterns.begin(), whitelist_patterns.end(),
        [&firstPartyOrigin](const std::pair<URLPattern, std::vector<URLPattern>> &info) {
          return info.first.MatchesURL(firstPartyOrigin);
        });

  if (i == whitelist_patterns.end()) {
    return false;
  }

  std::vector<URLPattern> &exceptions = i->second;
  return std::any_of(exceptions.begin(), exceptions.end(),
      [&subresourceUrl](const URLPattern& pattern) {
        return pattern.MatchesURL(subresourceUrl);
      });
}
```
And it relies on the empty `firstPartyOrigin` which is not good and is the subject of @diracdeltas notice.

**Comment by AlexeyBarabash** [comment]
Force-pushed the fix, created a new issue about empty url https://github.com/brave/browser-android-tabs/issues/963, it needs more investigating and probably redo by brave-core way.

File: chrome/browser/net/blockers/shield_exceptions.cc [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace net {
+namespace blockers {
+
+bool IsWhitelistedCookieExeption(const GURL& first_party_url,
+    const GURL& subresource_url) {
+  // Note that there's already an exception for TLD+1, so don't add those here.
+  // Check with the security team before adding exceptions.
+
+  if (first_party_url.is_empty()) {
```
Comment (diracdeltas) [inline chrome/browser/net/blockers/shield_exceptions.cc]: is this necessary? it seems bad privacy-wise if HTC cookies are allowed whenever the first party url is empty.

Comment (diracdeltas) [reply chrome/browser/net/blockers/shield_exceptions.cc]: i'm assuming it is based on https://github.com/brave/browser-android-tabs/issues/963

**Comment by AlexeyBarabash** [comment]
Updated PR. Got rid of empty url by brave-core method, updated whitelist.

**Comment by diracdeltas** [comment]
lgtm

**Comment by AlexeyBarabash** [comment]
Updated PR, simplified exception according the notice.

**Comment by AlexeyBarabash** [comment]
Created a separate issue for HTC account exception https://github.com/brave/browser-android-tabs/issues/964 .

**Pull request merged**